File: agentmail-note-taker/typescript/src/actions.ts
/**
 * actions.csv — flat ledger of action items extracted across notes.
 */

import { existsSync, appendFileSync, readFileSync, writeFileSync } from "node:fs";
import { createHash } from "node:crypto";

const FILE = "actions.csv";
const HEADER = ["id", "note_path", "owner", "task", "deadline", "urgency", "status", "created"];

function ensureHeader(): void {
  if (!existsSync(FILE)) writeFileSync(FILE, HEADER.join(",") + "\n");
}

function escapeCsv(v: string): string {
  if (v.includes(",") || v.includes('"') || v.includes("\n")) {
    return `"${v.replace(/"/g, '""')}"`;
  }
  return v;
}

function parseLine(line: string): string[] {
  const out: string[] = [];
  let cur = "", inQuote = false;
  for (let i = 0; i < line.length; i++) {
    const c = line[i];
    if (inQuote) {
      if (c === '"' && line[i + 1] === '"') { cur += '"'; i++; }
      else if (c === '"') inQuote = false;
      else cur += c;
    } else {
      if (c === '"') inQuote = true;
      else if (c === ",") { out.push(cur); cur = ""; }
      else cur += c;
    }
  }
  out.push(cur);
  return out;
}

function actionId(notePath: string, owner: string, task: string): string {
  return createHash("sha1").update(`${notePath}|${owner}|${task}`).digest("hex").slice(0, 12);
}

export interface ActionRow {
  id: string;
  note_path: string;
  owner: string;
  task: string;
  deadline: string;
  urgency: string;
  status: string;
  created: string;
}

function readAll(): ActionRow[] {
  if (!existsSync(FILE)) return [];
  const lines = readFileSync(FILE, "utf-8").split("\n").filter(l => l.length > 0);
  if (lines.length < 2) return [];
  const headers = parseLine(lines[0]);
  return lines.slice(1).map(line => {
    const fields = parseLine(line);
    const row: any = {};
    headers.forEach((h, i) => row[h] = fields[i] ?? "");
    return row as ActionRow;
  });
}

function writeAll(rows: ActionRow[]): void {
  ensureHeader();
  const lines = [HEADER.join(",")];
  for (const r of rows) {
    lines.push(HEADER.map(h => escapeCsv(String((r as any)[h] ?? ""))).join(","));
  }
  writeFileSync(FILE, lines.join("\n") + "\n");
}

export function appendFromNote(
  notePath: string,
  actionItems: Array<{ owner?: string; task?: string; deadline?: string; urgency?: string }>,
): ActionRow[] {
  ensureHeader();
  const existing = readAll();
  const existingIds = new Set(existing.map(r => r.id));
  const newRows: ActionRow[] = [];

  for (const ai of actionItems) {
    const owner = (ai.owner || "").trim();
    const task = (ai.task || "").trim();
    if (!task) continue;
    const id = actionId(notePath, owner, task);
    if (existingIds.has(id)) continue;
    const row: ActionRow = {
      id, note_path: notePath, owner, task,
      deadline: (ai.deadline || "").trim(),
      urgency: (ai.urgency || "").trim(),
      status: "open",
      created: new Date().toISOString().slice(0, 19) + "+00:00",
    };
    newRows.push(row);
    existingIds.add(id);
  }

  if (newRows.length) {
    const csvLines = newRows.map(r =>
      HEADER.map(h => escapeCsv(String((r as any)[h] ?? ""))).join(",")
    );
    appendFileSync(FILE, csvLines.join("\n") + "\n");
  }
  return newRows;
}

export function markDoneForNote(notePath: string): number {
  const rows = readAll();
  let n = 0;
  for (const r of rows) {
    if (r.note_path === notePath && r.status === "open") {
      r.status = "done";
      n++;
    }
  }
  if (n) writeAll(rows);
  return n;
}

export function listOpen(): ActionRow[] {
  return readAll().filter(r => r.status === "open");
}

export function isOverdue(row: ActionRow, today: Date): boolean {
  if (!row.deadline) return false;
  const d = new Date(row.deadline);
  if (isNaN(d.getTime())) return false;
  return d < new Date(today.toISOString().slice(0, 10));
}

export function hoursUntil(row: ActionRow, now: Date): number | null {
  if (!row.deadline) return null;
  let d = new Date(row.deadline);
  if (isNaN(d.getTime())) return null;
  // Date-only deadlines → end of day
  if (d.getUTCHours() === 0 && d.getUTCMinutes() === 0) {
    d = new Date(d.getTime() + (23 * 60 + 59) * 60 * 1000);
  }
  return (d.getTime() - now.getTime()) / 3_600_000;
}

File: agentmail-note-taker/python/actions.py
"""
actions.csv — flat ledger of all action items extracted across notes.

Columns:
  id, note_path, owner, task, deadline, urgency, status, created

`id` is a stable per-row hash so we can update status / dedup reminders.
`status` is one of: open, done, snoozed.
"""

import csv
import hashlib
from datetime import datetime, timezone
from pathlib import Path

ACTIONS_FILE = Path("actions.csv")
HEADER = ["id", "note_path", "owner", "task", "deadline", "urgency", "status", "created"]


def _ensure_header() -> None:
    if not ACTIONS_FILE.exists():
        with ACTIONS_FILE.open("w", newline="") as f:
            csv.writer(f).writerow(HEADER)


def _action_id(note_path: str, owner: str, task: str) -> str:
    return hashlib.sha1(f"{note_path}|{owner}|{task}".encode()).hexdigest()[:12]


def _read_all() -> list[dict]:
    if not ACTIONS_FILE.exists():
        return []
    with ACTIONS_FILE.open("r", newline="") as f:
        return list(csv.DictReader(f))


def _write_all(rows: list[dict]) -> None:
    _ensure_header()
    with ACTIONS_FILE.open("w", newline="") as f:
        w = csv.DictWriter(f, fieldnames=HEADER)
        w.writeheader()
        for r in rows:
            w.writerow({k: r.get(k, "") for k in HEADER})


def append_from_note(note_path: str, action_items: list[dict]) -> list[dict]:
    """Append actions; skip duplicates (same note_path + owner + task).
    Returns the list of newly-added rows (with ids assigned)."""
    _ensure_header()
    existing = _read_all()
    existing_ids = {r["id"] for r in existing}

    new_rows: list[dict] = []
    for ai in action_items:
        owner = (ai.get("owner") or "").strip()
        task = (ai.get("task") or "").strip()
        if not task:
            continue
        aid = _action_id(note_path, owner, task)
        if aid in existing_ids:
            continue
        row = {
            "id": aid,
            "note_path": note_path,
            "owner": owner,
            "task": task,
            "deadline": (ai.get("deadline") or "").strip(),
            "urgency": (ai.get("urgency") or "").strip(),
            "status": "open",
            "created": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        }
        new_rows.append(row)
        existing_ids.add(aid)

    if new_rows:
        with ACTIONS_FILE.open("a", newline="") as f:
            w = csv.DictWriter(f, fieldnames=HEADER)
            for r in new_rows:
                w.writerow(r)
    return new_rows


def mark_done_for_note(note_path: str) -> int:
    """Mark all open actions for the given note as done. Return count."""
    rows = _read_all()
    n = 0
    for r in rows:
        if r["note_path"] == note_path and r["status"] == "open":
            r["status"] = "done"
            n += 1
    if n:
        _write_all(rows)
    return n


def list_open() -> list[dict]:
    return [r for r in _read_all() if r["status"] == "open"]


def list_for_note(note_path: str) -> list[dict]:
    return [r for r in _read_all() if r["note_path"] == note_path]


def is_overdue(row: dict, today: datetime) -> bool:
    if not row.get("deadline"):
        return False
    try:
        d = datetime.fromisoformat(row["deadline"]).date()
    except Exception:
        return False
    return d < today.date()


def hours_until(row: dict, now: datetime) -> float | None:
    if not row.get("deadline"):
        return None
    try:
        d = datetime.fromisoformat(row["deadline"])
        # Treat date-only deadlines as end-of-day
        if d.time().hour == 0 and d.time().minute == 0:
            d = d.replace(hour=23, minute=59)
    except Exception:
        return None
    return (d - now.replace(tzinfo=None)).total_seconds() / 3600.0
